integration/rpctest/btcd.go

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2017 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package rpctest
+
+import (
+	"fmt"
+	"go/build"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"sync"
+)
+
+var (
+	// btcdCompileMtx guards access to the btcd path so that it is only compiled
+	// once.
+	btcdCompileMtx sync.Mutex
+
+	// _btcdFilepath is the path to the btcd executable. This is the empty
+	// string until btcd is compiled.
+	_btcdFilepath string
+)
+
+// btcdExecutablePath returns a path to the btcd executable to be used by
+// rpctests. To ensure the code tests against the most up-to-date version of
+// btcd, this method compiles btcd the first time it is called. After that, the
+// generated binary is used for subsequent test harnesses. The executable file
+// is not cleanup up, but since it lives at a static path in a temp directory,
+// it is not a big deal.
+func btcdExecutablePath() (string, error) {
+	btcdCompileMtx.Lock()
+
+	// If btcd has already been compiled, just use that.
+	if len(_btcdFilepath) != 0 {
+		btcdCompileMtx.Unlock()
+		return _btcdFilepath, nil
+	}
+
+	defer btcdCompileMtx.Unlock()
```

Comment:
Minor nitpick, but might as well just put the defer after the lock and remove the explicit unlock above.  This is only test code, so saving the defer overhead doesn't seem worth the readability hit to me.  I actually missed this on my first read through and though the variable was used below without the mutex!